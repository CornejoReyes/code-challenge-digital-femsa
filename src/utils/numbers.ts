export function formatNumberAsQuantity(value: number): string {
  const formatter = new Intl.NumberFormat('es-MX', { minimumFractionDigits: 2, maximumFractionDigits: 2 });
  return formatter.format(value);
}
