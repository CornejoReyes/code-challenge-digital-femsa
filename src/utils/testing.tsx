import { render, RenderOptions } from '@testing-library/react-native';
import { ReactElement } from 'react';
import Providers from '../Providers';

function customRender(ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) {
  return render(ui, { wrapper: Providers, ...options });
}

export * from '@testing-library/react-native';
export { customRender as render };
