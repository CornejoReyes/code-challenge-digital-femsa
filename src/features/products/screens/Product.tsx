import { NativeStackScreenProps } from '@react-navigation/native-stack';
import { format, parseISO } from 'date-fns';
import { es } from 'date-fns/locale';
import { AspectRatio, Box, Button, Factory, StatusBar, Text, useTheme } from 'native-base';
import FastImage from 'react-native-fast-image';
import { RootStackParamList } from '../../../navigation/types';
import { formatNumberAsQuantity } from '../../../utils/numbers';

const Image = Factory(FastImage);

export interface Props extends NativeStackScreenProps<RootStackParamList, 'ProductDetails'> {}

export default function Product({ route, navigation }: Props) {
  const theme = useTheme();
  const product = route.params.product;
  const parsedDate = parseISO(product.createdAt);
  return (
    <>
      <StatusBar backgroundColor={theme.colors.secondary} barStyle="dark-content" />
      <Box flex="1" safeAreaBottom>
        <Box bg="secondary" h="15%" justifyContent="flex-end" p="4">
          <Text fontWeight="bold" fontSize="lg">
            {product.product}
          </Text>
        </Box>
        <Box flex="1" p="4">
          <AspectRatio ratio={1} mb="4">
            <Box bg="white" rounded="xl" shadow="4" p="8">
              <Image w="full" h="full" source={{ uri: product.image }} />
            </Box>
          </AspectRatio>
          <Text fontSize="xs" fontWeight="bold" color="gray.500" mb="4">
            Detalles del producto:
          </Text>
          <Text fontWeight="bold" mb="4" accessibilityLabel="product date">
            Comprado el {format(parsedDate, "dd 'de' MMMM ',' yyyy", { locale: es })}
          </Text>
          <Text fontSize="xs" fontWeight="bold" color="gray.500" mb="4" accessibilityLabel="product transaction label">
            Con esta compra {product.is_redemption ? 'canjeaste' : 'acumulaste'}:
          </Text>
          <Text fontWeight="bold" mb="4" accessibilityLabel="product transaction">
            {formatNumberAsQuantity(product.points)} puntos
          </Text>
        </Box>
        <Button variant="main" m="4" onPress={navigation.goBack} accessibilityLabel="cta">
          Aceptar
        </Button>
      </Box>
    </>
  );
}
