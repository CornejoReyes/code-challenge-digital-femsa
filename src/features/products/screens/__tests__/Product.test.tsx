import { fireEvent, render, screen } from '../../../../utils/testing';
import Product from '../Product';

const mockProduct = {
  createdAt: '2022-12-09T10:20:00.909Z',
  product: 'Product item 1',
  points: 100,
  image: 'image',
  is_redemption: false,
  id: '1',
};

describe('<Product />', () => {
  it('renders a product correctly', () => {
    const route = {
      params: {
        product: mockProduct,
      },
    };
    render(<Product navigation={{} as any} route={route as any} />);

    const date = screen.getByLabelText('product date');
    const transactionLabel = screen.getByLabelText('product transaction label');
    const transaction = screen.getByLabelText('product transaction');

    expect(screen.toJSON()).toMatchSnapshot();
    expect(date).toHaveTextContent('Comprado el 09 de diciembre , 2022');
    expect(transactionLabel).toHaveTextContent('Con esta compra acumulaste');
    expect(transaction).toHaveTextContent('100.00 puntos');
  });

  it('renders a product with redemption correctly', () => {
    const productWithRedemption = {
      ...mockProduct,
      is_redemption: true,
    };
    const route = {
      params: {
        product: productWithRedemption,
      },
    };
    render(<Product navigation={{} as any} route={route as any} />);

    const transactionLabel = screen.getByLabelText('product transaction label');

    expect(screen.toJSON()).toMatchSnapshot();
    expect(transactionLabel).toHaveTextContent('Con esta compra canjeaste');
  });

  it('goes back when button is pressed', () => {
    const mockBack = jest.fn();
    const navigation = {
      goBack: mockBack,
    };
    const route = {
      params: {
        product: mockProduct,
      },
    };
    render(<Product navigation={navigation as any} route={route as any} />);

    const button = screen.getByLabelText('cta');

    fireEvent.press(button);

    expect(mockBack).toHaveBeenCalled();
  });
});
