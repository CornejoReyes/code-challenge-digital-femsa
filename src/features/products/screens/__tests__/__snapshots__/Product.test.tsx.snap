// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product /> renders a product correctly 1`] = `
<View
  dataSet={{}}
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    dataSet={{}}
    style={
      {
        "backgroundColor": "#CFD6FF",
        "height": "15%",
        "justifyContent": "flex-end",
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
      }
    }
  >
    <Text
      dataSet={{}}
      style={
        {
          "backgroundColor": undefined,
          "color": "#171717",
          "fontFamily": undefined,
          "fontSize": 18,
          "fontStyle": "normal",
          "fontWeight": "700",
          "letterSpacing": 0,
          "lineHeight": 27,
          "textDecorationLine": undefined,
        }
      }
    >
      Product item 1
    </Text>
  </View>
  <View
    dataSet={{}}
    style={
      {
        "flex": 1,
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
      }
    }
  >
    <View
      dataSet={{}}
      style={
        [
          {
            "marginBottom": 16,
          },
          {
            "aspectRatio": 1,
            "style": undefined,
          },
        ]
      }
    >
      <View
        dataSet={{}}
        style={
          [
            {
              "backgroundColor": "#FFFFFF",
              "borderRadius": 12,
              "elevation": 5,
              "paddingBottom": 32,
              "paddingLeft": 32,
              "paddingRight": 32,
              "paddingTop": 32,
              "shadowColor": "#000000",
              "shadowOffset": {
                "height": 2,
                "width": 0,
              },
              "shadowOpacity": 0.25,
              "shadowRadius": 3.84,
            },
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
          ]
        }
      >
        <View
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "height": "100%",
                "width": "100%",
              },
            ]
          }
        >
          <FastImageView
            dataSet={{}}
            defaultSource={null}
            resizeMode="cover"
            source={
              {
                "uri": "image",
              }
            }
            style={
              {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          />
        </View>
      </View>
    </View>
    <Text
      dataSet={{}}
      style={
        {
          "backgroundColor": undefined,
          "color": "#71717a",
          "fontFamily": undefined,
          "fontSize": 12,
          "fontStyle": "normal",
          "fontWeight": "700",
          "letterSpacing": 0,
          "lineHeight": 18,
          "marginBottom": 16,
          "textDecorationLine": undefined,
        }
      }
    >
      Detalles del producto:
    </Text>
    <Text
      accessibilityLabel="product date"
      dataSet={{}}
      style={
        {
          "backgroundColor": undefined,
          "color": "#171717",
          "fontFamily": undefined,
          "fontSize": 14,
          "fontStyle": "normal",
          "fontWeight": "700",
          "letterSpacing": 0,
          "lineHeight": 21,
          "marginBottom": 16,
          "textDecorationLine": undefined,
        }
      }
    >
      Comprado el 
      09 de diciembre , 2022
    </Text>
    <Text
      accessibilityLabel="product transaction label"
      dataSet={{}}
      style={
        {
          "backgroundColor": undefined,
          "color": "#71717a",
          "fontFamily": undefined,
          "fontSize": 12,
          "fontStyle": "normal",
          "fontWeight": "700",
          "letterSpacing": 0,
          "lineHeight": 18,
          "marginBottom": 16,
          "textDecorationLine": undefined,
        }
      }
    >
      Con esta compra 
      acumulaste
      :
    </Text>
    <Text
      accessibilityLabel="product transaction"
      dataSet={{}}
      style={
        {
          "backgroundColor": undefined,
          "color": "#171717",
          "fontFamily": undefined,
          "fontSize": 14,
          "fontStyle": "normal",
          "fontWeight": "700",
          "letterSpacing": 0,
          "lineHeight": 21,
          "marginBottom": 16,
          "textDecorationLine": undefined,
        }
      }
    >
      100.00
       puntos
    </Text>
  </View>
  <View
    accessibilityLabel="cta"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    colorScheme="primary"
    dataSet={{}}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#334FFA",
        "borderRadius": 8,
        "elevation": 5,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 16,
        "marginLeft": 16,
        "marginRight": 16,
        "marginTop": 16,
        "paddingBottom": 10,
        "paddingLeft": 12,
        "paddingRight": 12,
        "paddingTop": 10,
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 3.84,
      }
    }
  >
    <View
      dataSet={{}}
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
      test={true}
    >
      <View
        dataSet={{}}
        style={{}}
      >
        <Text
          dataSet={{}}
          style={
            {
              "backgroundColor": undefined,
              "color": "#FFFFFF",
              "fontFamily": undefined,
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "700",
              "letterSpacing": 0,
              "lineHeight": 21,
              "textDecorationLine": undefined,
            }
          }
        >
          Aceptar
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`<Product /> renders a product with redemption correctly 1`] = `
<View
  dataSet={{}}
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    dataSet={{}}
    style={
      {
        "backgroundColor": "#CFD6FF",
        "height": "15%",
        "justifyContent": "flex-end",
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
      }
    }
  >
    <Text
      dataSet={{}}
      style={
        {
          "backgroundColor": undefined,
          "color": "#171717",
          "fontFamily": undefined,
          "fontSize": 18,
          "fontStyle": "normal",
          "fontWeight": "700",
          "letterSpacing": 0,
          "lineHeight": 27,
          "textDecorationLine": undefined,
        }
      }
    >
      Product item 1
    </Text>
  </View>
  <View
    dataSet={{}}
    style={
      {
        "flex": 1,
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
      }
    }
  >
    <View
      dataSet={{}}
      style={
        [
          {
            "marginBottom": 16,
          },
          {
            "aspectRatio": 1,
            "style": undefined,
          },
        ]
      }
    >
      <View
        dataSet={{}}
        style={
          [
            {
              "backgroundColor": "#FFFFFF",
              "borderRadius": 12,
              "elevation": 5,
              "paddingBottom": 32,
              "paddingLeft": 32,
              "paddingRight": 32,
              "paddingTop": 32,
              "shadowColor": "#000000",
              "shadowOffset": {
                "height": 2,
                "width": 0,
              },
              "shadowOpacity": 0.25,
              "shadowRadius": 3.84,
            },
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
          ]
        }
      >
        <View
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "height": "100%",
                "width": "100%",
              },
            ]
          }
        >
          <FastImageView
            dataSet={{}}
            defaultSource={null}
            resizeMode="cover"
            source={
              {
                "uri": "image",
              }
            }
            style={
              {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          />
        </View>
      </View>
    </View>
    <Text
      dataSet={{}}
      style={
        {
          "backgroundColor": undefined,
          "color": "#71717a",
          "fontFamily": undefined,
          "fontSize": 12,
          "fontStyle": "normal",
          "fontWeight": "700",
          "letterSpacing": 0,
          "lineHeight": 18,
          "marginBottom": 16,
          "textDecorationLine": undefined,
        }
      }
    >
      Detalles del producto:
    </Text>
    <Text
      accessibilityLabel="product date"
      dataSet={{}}
      style={
        {
          "backgroundColor": undefined,
          "color": "#171717",
          "fontFamily": undefined,
          "fontSize": 14,
          "fontStyle": "normal",
          "fontWeight": "700",
          "letterSpacing": 0,
          "lineHeight": 21,
          "marginBottom": 16,
          "textDecorationLine": undefined,
        }
      }
    >
      Comprado el 
      09 de diciembre , 2022
    </Text>
    <Text
      accessibilityLabel="product transaction label"
      dataSet={{}}
      style={
        {
          "backgroundColor": undefined,
          "color": "#71717a",
          "fontFamily": undefined,
          "fontSize": 12,
          "fontStyle": "normal",
          "fontWeight": "700",
          "letterSpacing": 0,
          "lineHeight": 18,
          "marginBottom": 16,
          "textDecorationLine": undefined,
        }
      }
    >
      Con esta compra 
      canjeaste
      :
    </Text>
    <Text
      accessibilityLabel="product transaction"
      dataSet={{}}
      style={
        {
          "backgroundColor": undefined,
          "color": "#171717",
          "fontFamily": undefined,
          "fontSize": 14,
          "fontStyle": "normal",
          "fontWeight": "700",
          "letterSpacing": 0,
          "lineHeight": 21,
          "marginBottom": 16,
          "textDecorationLine": undefined,
        }
      }
    >
      100.00
       puntos
    </Text>
  </View>
  <View
    accessibilityLabel="cta"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    colorScheme="primary"
    dataSet={{}}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#334FFA",
        "borderRadius": 8,
        "elevation": 5,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 16,
        "marginLeft": 16,
        "marginRight": 16,
        "marginTop": 16,
        "paddingBottom": 10,
        "paddingLeft": 12,
        "paddingRight": 12,
        "paddingTop": 10,
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 3.84,
      }
    }
  >
    <View
      dataSet={{}}
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
      test={true}
    >
      <View
        dataSet={{}}
        style={{}}
      >
        <Text
          dataSet={{}}
          style={
            {
              "backgroundColor": undefined,
              "color": "#FFFFFF",
              "fontFamily": undefined,
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "700",
              "letterSpacing": 0,
              "lineHeight": 21,
              "textDecorationLine": undefined,
            }
          }
        >
          Aceptar
        </Text>
      </View>
    </View>
  </View>
</View>
`;
